src/NHibernate/Tuple/Entity/PocoEntityTuplizer.cs

```diff
@@ -66,6 +68,8 @@ public PocoEntityTuplizer(EntityMetamodel entityMetamodel, PersistentClass mappe
 				if (property.UnwrapProxy)
 					unwrapProxyPropertyNames.Add(property.Name);
 			}
+
+			isBytecodeProviderImpl = Cfg.Environment.BytecodeProvider is BytecodeProviderImpl;
```

Comment:
It should be moved into `SetReflectionOptimizer`, because nothing guarantees the byte code provider will be the same on deserialization. By the way the test could be on the `optimizer.AccessOptimizer` concrete implementation instead, which would be more straightforward.

Reply:
Moved to into `SetReflectionOptimizer`. 

> By the way the test could be on the optimizer.AccessOptimizer concrete implementation instead

Is it not enough as the user can use the obsolete constructor that does not have the specialized getter/setter and it would throw a NRE on (Set\Get)SpecializedPropertyValue. We have to use the same condition as in the `GetReflectionOptimizer` extension method in order to avoid that.